Readme.md

# Js inheritance test project

Description

A simple test project to try out javascript inheritance.

## License 

(The MIT License)

Copyright (c) 2009 Gewalli &lt;gewalli (at) gmail&gt;

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

History.md
0.0.1 / 2011-05-13
------------------

* Initial release

Some tests of javascript inheritance.

0.0.2 / 2011-09-04
------------------

* Test of something akin to override 

0.0.3 / 2011-09-04
------------------

* JsDoc

Generate documentation based on javascript. Mostly to try out jsdoc-toolkit.

0.0.4 / 2015-04-07
------------------

Trying out [Object.create prototype](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Inheritance_and_the_prototype_chain)


spec/node.js

require('./lib/jspec')
require('./unit/spec.helper')
require('../lib/application')

JSpec
  .exec('spec/unit/spec.js')
  .run({ reporter: JSpec.reporters.Terminal, fixturePath: 'spec/fixtures', failuresOnly: true })
  .report()


spec.js

describe 'standard inheritance'
	before_each
		var desc = this;
		desc.FurryAnimal = function (name){
			var self = this;
			this.m_name = name;
			this.sayName = function(){return self.m_name;};
		};
		desc.Feline = function (name){
			var self = this;
			desc.FurryAnimal.call(this,name);
		};
		desc.Cat = function(name){
			var self = this;
			desc.Feline.apply(this,arguments);
		};
	end
	describe 'rat'
		before_each
			this.animal = new this.FurryAnimal("rat");
		end
		it "animal should say rat"
			this.animal.sayName().should.be("rat");
		end
	end
	describe 'lion'
		before_each
			this.lion = new this.Feline("lion");
		end
		it "should say lion"
			this.lion.sayName().should.be("lion")
		end
		it "should despite redefining 'this' say lion"
			this.lion.sayName.call({m_name:"red"}).should.be("lion");
		end
	end
	describe 'cat'
		before_each
			this.cat = new this.Cat("missan");
		end
		it "should say missan"
			this.cat.sayName().should.be("missan");
		end
	end
end
describe 'prototype'
	describe 'animal'
		before_each
			var desc = this;
			desc.Animal = function (name){
				var self = this;
				this.m_name = name;
				this.sayName = function(){return self.m_name;};
			};
		end
		it "protect"
			var desc = this;
			var animal = app.protectThis(new desc.Animal("animal"));
			animal.sayName.call({m_name:"name"}).should.be("animal");
		end
	end
	describe 'self'
		before_each
			var desc = this;
			desc.SpineLess = function(name){
				//this.m_name = name;
				this.getSelf().m_name = name;
			};
			desc.SpineLess.prototype = new (function(){
				var self = this;//Since were working in a prototype, this isnt safe.
				this.m_name = "none";
				this.sayName = function(){return self.m_name;};
				this.getSelf = function(){return self;};//The danger of prototype inheritance
			})();
		end
	
		/*app.protectThis*/
		describe 'mollusc'
			before_each
				this.mollusc = new this.SpineLess("mollusc");
				this.annelid = new this.SpineLess("annelid");
			end
			it "mollusc should say annelid!"
				this.mollusc.sayName().should.be("annelid");
			end
			it "annelid should say annelid"
				this.annelid.sayName().should.be("annelid");
			end
			it "mollusc should be able to change name of annelid"
				this.mollusc.getSelf().m_name = "mollusc";
				this.annelid.sayName().should.be("mollusc");
			end
		end
	end
end
/*
describe 'mootools inheritance'
	describe 'animal'
		before_each
			var desc = this;
			desc.MooAnimal = new Class({
				initialize: function(age){
					this.age = age;
					this.name = "";
				},
				getAge: function(){
					return this.age;
				},
				setName: function(name){
					this.name = name
				}
			});
		end
		it "should have age"
			var desc = this;
			print(desc.MooAnimal);
			var myAnimal = new desc.MooAnimal(22);
			myAnimal.age.should.be(22);
			myAnimal.getAge().should.be(22);
		end
		it "should be able to set name"
			var desc = this;
			print(desc.MooAnimal);
			var myAnimal = new desc.MooAnimal(22);
			myAnimal.setName("flux");
			myAnimal.name.should.be("flux");
		end
		describe 'cat'
			before_each
				var desc = this;
				desc.MooCat = new Class({
					Extends: desc.MooAnimal,
					initialize: function(name, age){
						this.parent(age); //will call initalize of Animal
						this.name = name;
					}
				});
			end
			it 'should be named micia'
				print(this.MooCat);
				var myCat = new this.MooCat('Micia', 20);
				myCat.should.not.be_undefined();
				for(x in myCat){ print (""+x);}
				myCat.name.should.be("Micia"); //Alerts 'Micia'.
				myCat.age.should.be(20); //Alerts 20.

			end
		end
	end
end*/

describe 'app'
	before_each
		var desc = this;
		desc.Animal = function(){
			this.breathes = true;
			this.walk = false;
			this.dowalk = function(){
				this.walk = true;
			};
		};
	end
	describe 'Animal'
		it 'Animal class should exist'
			this.Animal.should.not.be_undefined
		end
		it "Animal class should be constructable"
			var a = new this.Animal();
			a.should.not.be_undefined();
			(a instanceof this.Animal).should.be_true();
		end
		it "Animal class should require new"
			try {
				var a = this.Animal();
			}
			catch (e) {
				(e).should.not.be_undefined();
				(e instanceof TypeError).should.be_true();
			}
			(a).should.be_undefined();
		end
		it "Animal object should breath"
			var a = new this.Animal();
			(a.breathes).should.be_true();
		end
		it "Animal object should be able to walk"
			var a = new this.Animal();
			a.dowalk();
			(a.walk).should.be_true();
		end
		describe "feline"
			before_each
				var desc = this;
				desc.Feline = function(){
					desc.Animal.call(this);
					this.claws = true;
					this.furry = true;
					this.says = function() {
						console.log ('GRRRRR');
					};
					this.declaw = function() {
						this.claws = false;
					};
				};
			end
			it "Feline class should exist"
				(this.Feline).should.not.be_null();
			end
			it "Feline class should be constructable"
				var f = new this.Feline();
				(this.Feline).should.not.be_undefined();
				(f instanceof this.Feline).should.be_true();
			end
			it "Feline object should breath"
				var f = new this.Feline();
				(f.breathes).should.be_true();
			end
			it "Feline object should have claws and be furry"
				var f = new this.Feline();
				(f.claws).should.be_true();
				(f.furry).should.be_true();
			end
			it "Should be able to declaw a feline object"
				var f = new this.Feline();
				var f2 = new this.Feline();
				f.declaw();
				(f.claws).should.be_false();
				(f2.claws).should.be_true();
			end
			it "Should be able to declaw a feline object despite redefining this"
				var f = app.protectThis(new this.Feline());
				f.declaw.call({claws:true});
				(f.claws).should.be_false();
			end
			it "Feline object should be able to walk"
				var a = new this.Feline();
				a.dowalk();
				(a.walk).should.be_true();
			end
		end
	end
end

lib/application.js
/** @namespace */
app = {};
/**
* Takes an iterable and transforms it into an array.
*/
app.toArray = function (iterable) {
		return Array.prototype.slice.call(iterable);
	};

/**
* Inspired by Cobra.Class.method(callable, self)
* It will wrap a function to give it a new _this_ context.
*/
app.methodize = function (/** the function to wrap inside a new function call to give it a new this context.*/ callable, /** self is the python verb for _this_ class.*/ self) {
	
	function method () {
		var args = app.toArray(arguments);
		return callable.apply(self, args);
	}
	method.callable = callable;
	return method;
};

/**
* This function takes an object and wraps each of it's functions in a call that makes sure that 
* this will be bound to the object.
*/
app.protectThis = function(object){
  var key;
	for (key in object) {
		if (object.hasOwnProperty(key)) {
			var member = object[key];
	
			if (typeof member == 'function') {
				object[key] = app.methodize(member, object);
			}
		}
	}
	return object;
};

/**
* This function takes an object, a name to the function to override and a new function that overrides the old.
* In order to pass on execution to the old function you use this(args) inside the new function. 
*/
app.override = function(self, /**The name of the function to override.*/ key, /**The child function overriding the parent.*/ child){
    var parent = self[key];
    /** @private */
    self[key] = function(){
        return child.apply(parent,arguments);
    };
};

app.getName = function(self){
    var funcNameRegex = /function (.{1,})\(/;
    var results = (funcNameRegex).exec((self).constructor.toString());
    return (results && results.length > 1) ? results[1] : "";
};



jsdoc/app/lib/JSDOC.js
/**
	@overview
	@date $Date: 2010-07-03 09:55:05 +0200 (Lör, 03 Jul 2010) $
	@version $Revision: 846 $ 
	@location $HeadURL: http://jsdoc-toolkit.googlecode.com/svn/trunk/jsdoc-toolkit/app/lib/JSDOC.js $
	@name JSDOC.js
 */

/**
	This is the main container for the JSDOC application.
	@namespace
*/
JSDOC = {
};

/**
	@requires Opt
 */
if (typeof arguments == "undefined") arguments = [];
JSDOC.opt = Opt.get(
	arguments, 
	{
		a: "allfunctions", 
		c: "conf",
		d: "directory",
		"D[]": "define",
		e: "encoding",
		"E[]": "exclude",
		h: "help",
		m: "multiple",
		n: "nocode",
		o: "out",
		p: "private",
		P: "plugins",
		q: "quiet",
		r: "recurse",
		S: "securemodules",
		s: "suppress",
		t: "template",
		T: "testmode",
		u: "unique",
		v: "verbose",
		x: "ext"
	}
);

/** The current version string of this application. */
JSDOC.VERSION = "2.4.0";

/** Print out usage information and quit. */
JSDOC.usage = function() {
	print("USAGE: java -jar jsrun.jar app/run.js [OPTIONS] <SRC_DIR> <SRC_FILE> ...");
	print("");
	print("OPTIONS:");
	print("  -a or --allfunctions\n          Include all functions, even undocumented ones.\n");
	print("  -c or --conf\n          Load a configuration file.\n");
	print("  -d=<PATH> or --directory=<PATH>\n          Output to this directory (defaults to \"out\").\n");
	print("  -D=\"myVar:My value\" or --define=\"myVar:My value\"\n          Multiple. Define a variable, available in JsDoc as JSDOC.opt.D.myVar.\n");
	print("  -e=<ENCODING> or --encoding=<ENCODING>\n          Use this encoding to read and write files.\n");
	print("  -E=\"REGEX\" or --exclude=\"REGEX\"\n          Multiple. Exclude files based on the supplied regex.\n");
	print("  -h or --help\n          Show this message and exit.\n");
	print("  -m or --multiples\n          Don't warn about symbols being documented more than once.\n");
	print("  -n or --nocode\n          Ignore all code, only document comments with @name tags.\n");
	print("  -o=<PATH> or --out=<PATH>\n          Print log messages to a file (defaults to stdout).\n");
	print("  -p or --private\n          Include symbols tagged as private, underscored and inner symbols.\n");
	print("  -P or --plugins\n          Load additional plugins from the specified folder.\n");
	print("  -q or --quiet\n          Do not output any messages, not even warnings.\n");
	print("  -r=<DEPTH> or --recurse=<DEPTH>\n          Descend into src directories.\n");
	print("  -s or --suppress\n          Suppress source code output.\n");
	print("  -S or --securemodules\n          Use Secure Modules mode to parse source code.\n");
	print("  -t=<PATH> or --template=<PATH>\n          Required. Use this template to format the output.\n");
	print("  -T or --test\n          Run all unit tests and exit.\n");
	print("  -u or --unique\n          Force file names to be unique, but not based on symbol names.\n");
	print("  -v or --verbose\n          Provide verbose feedback about what is happening.\n");
	print("  -x=<EXT>[,EXT]... or --ext=<EXT>[,EXT]...\n          Scan source files with the given extension/s (defaults to js).\n");
	
	quit();
}

/*t:
	plan(4, "Testing JSDOC namespace.");
	
	is(
		typeof JSDOC,
		"object",
		"JSDOC.usage is a function."
	);
	
	is(
		typeof JSDOC.VERSION,
		"string",
		"JSDOC.VERSION is a string."
	);
	
	is(
		typeof JSDOC.usage,
		"function",
		"JSDOC.usage is a function."
	);
	
	is(
		typeof JSDOC.opt,
		"object",
		"JSDOC.opt is a object."
	);
 */

if (this.IO) IO.includeDir("lib/JSDOC/");
